components/layout_2020/flow/root.rs

```diff
@@ -359,3 +375,72 @@ impl FragmentTree {
         .unwrap_or_else(Rect::zero)
     }
 }
+
+/// https://drafts.csswg.org/css-backgrounds/#root-background
+#[derive(Clone)]
+pub(crate) struct CanvasBackground {
+    pub style: Option<Arc<ComputedValues>>,
+
+    /// The element whose style we have above
+    pub from_element: OpaqueNode,
+
+    pub root_element: OpaqueNode,
+}
+
+impl CanvasBackground {
+    fn for_root_element<'dom>(context: &LayoutContext, root_element: impl NodeExt<'dom>) -> Self {
+        let root_style = root_element.style(context);
+
+        let mut style = root_style;
+        let mut from_element = root_element;
+
+        // https://drafts.csswg.org/css-backgrounds/#body-background
+        // “if the computed value of background-image on the root element is none
+        //  and its background-color is transparent”
+        if style.background_is_transparent() &&
+            // “For documents whose root element is an HTML `HTML` element
+            //  or an XHTML `html` element”
+            root_element.type_id() == LayoutNodeType::Element(LayoutElementType::HTMLHtmlElement) &&
+            // Don’t try to access styles for an unstyled subtree
+            !matches!(style.clone_display().into(), Display::None)
+        {
+            // “that element’s first HTML `BODY` or XHTML `body` child element”
+            if let Some(body) = iter_child_nodes(root_element).find(|child| {
+                child.is_element() &&
+                    child.type_id() ==
+                        LayoutNodeType::Element(LayoutElementType::HTMLBodyElement)
+            }) {
+                style = body.style(context);
+                from_element = body;
+            }
+        }
+
+        Self {
+            root_element: root_element.opaque(),
+            from_element: from_element.opaque(),
+
+            // “However, if no boxes are generated for the element
+            //  whose background would be used for the canvas
+            //  (for example, if the root element has display: none),
+            //  then the canvas background is transparent.”
+            style: if let Display::GeneratingBox(_) = style.clone_display().into() {
+                Some(style)
+            } else {
+                None
+            },
+        }
+    }
+}
+
+fn iter_child_nodes<'dom, Node>(parent: Node) -> impl Iterator<Item = Node>
```

Comment:
Is this duplicated from `components/layout_2020/dom_traversal.rs` on purpose?

Reply:
No, I meant to move it there and forget to remove the original :)